recipe-server/client/control/utils/handleError.js

```diff
@@ -0,0 +1,45 @@
+import { message } from 'antd';
+
+import APIClient from 'control/utils/api';
+
+
+export const ERR_MESSAGES = {
+  FORM_VALIDATION: 'Please correct the form items highlighted below.',
+  NO_INTERNET: 'Check your internet connection and try again.',
+  SERVER_FAILED: 'The server failed to respond. Please try again.',
+};
+
+// String to search for to determine if a fetch failed. Unlikely to change.
+const fetchFailureMessage = 'Failed to fetch';
+
+const defaultMethods = {
+  checkUserOnline: () => navigator.onLine,
+  notifyUser: errMsg => message.error(errMsg),
+};
+
+export default function handleError(context = 'Error!', error, methodOverrides = {}) {
+  const methods = { ...defaultMethods, ...methodOverrides };
+  let errMsg = '';
+
+  if (error) {
+    if (!methods.checkUserOnline()) {
+      errMsg = ERR_MESSAGES.NO_INTERNET;
+    } else if (error.message.indexOf(fetchFailureMessage) > -1) {
+      errMsg = ERR_MESSAGES.SERVER_FAILED;
+    } else if (error instanceof APIClient.APIError && error.data) {
+      errMsg = ERR_MESSAGES.FORM_VALIDATION;
```

Comment:
This isn't really accurate. You are pretty much ignoring the actual error from the server and assuming any time that the server responded with an error is a form validation error. I would like this to be much surer of itself before giving the form validation error message. It would be preferable to give uglier but more information dense answers instead of hiding details.

An example of this is a permission error. If you create a new user and don't give them any permissions, creating a recipe returns HTTP 403 and the body `{"detail":"You do not have permission to perform this action."}`. Compare to a form validation error: HTTP 400 with the body `{"name":["This field may not be blank."]}`. 

I think checking HTTP status codes is a necessary part of error handling that we don't do. A smaller change would be at least exposing the `detail` field if it exists.

Reply:
The `APIClient` already handles determining the `message` to return with the server error, I assumed it was safe for `handleError` to inherit it from there:
```
// control/utils/api.js
...
if (!response.ok) {
  let message;
  let data;
  let err;

  try {
    data = await response.json();
    message = data.detail || response.statusText;
  } catch (error) {
    message = error.message;
    err = error;
  }

  throw new APIClient.APIError(message, data, err);
}
...
```

I agree that checking against the `data` prop for the form validation errors is weak, I can find a way to strengthen that. I can work on `api.js` to inspect the status codes, as well.